File: WebApp/Scripts/Mobile/LambertConformalConic.js
//  Copyright 2012 Applied Geographics, Inc.  Licensed under the Apache License, Version 2.0 http://www.apache.org/licenses/LICENSE-2.0

var GPV = (function (gpv) {
  gpv.coordSys = function (params) {
    var halfPi = 1.5707963267948966192;
    var quarterPi = 0.7853981633974483096;
    var radiansPerDegree = 0.0174532925199432958;
    var degreesPerRadian = 57.295779513082320877;

    var metersPerUnit = params.units == "feet" ? 12.0 / 39.37 : 1;
    var unitsPerMeter = 1 / metersPerUnit;

    var centralMeridian = params.centralMeridian * radiansPerDegree;
    var originLatitude = params.originLatitude * radiansPerDegree;
    var standardParallel1 = params.standardParallel1 * radiansPerDegree;
    var standardParallel2 = params.standardParallel2 * radiansPerDegree;
    var falseEasting = params.falseEasting;
    var falseNorthing = params.falseNorthing;

    var spheroid = null;

    switch (params.spheroid) {
      case "Clarke1866": spheroid = { semiMajorAxis: 6378206.4, eccentricity: 0.0822718542230 }; break;
      case "GRS80": spheroid = { semiMajorAxis: 6378137, eccentricity: 0.0818191910435 }; break;
      case "WGS84": spheroid = { semiMajorAxis: 6378137, eccentricity: 0.0818191908426 }; break;
    }

    var m1 = getM(standardParallel1);
    var m2 = getM(standardParallel2);

    var t0 = getT(originLatitude);
    var t1 = getT(standardParallel1);
    var t2 = getT(standardParallel2);

    var pcN = (Math.log(m1) - Math.log(m2)) / (Math.log(t1) - Math.log(t2));
    var pcF = m1 / (pcN * Math.pow(t1, pcN));
    var pcR0 = getR(t0);

    function getM(lat) {
      var eSqr = spheroid.eccentricity * spheroid.eccentricity;
      return Math.cos(lat) / Math.sqrt(1 - (eSqr * Math.pow(Math.sin(lat), 2)));
    }

    function getR(t) {
      return spheroid.semiMajorAxis * pcF * Math.pow(t, pcN);
    }

    function getT(lat) {
      var sinLat = Math.sin(lat);
      return Math.tan(quarterPi - lat / 2) / Math.pow((1 - spheroid.eccentricity * sinLat) /
        (1 + spheroid.eccentricity * sinLat), spheroid.eccentricity / 2);
    }

    function toProjected(g) {
      var lon = g[0] * radiansPerDegree;
      var lat = g[1] * radiansPerDegree;

      var t = getT(lat);
      var r = getR(t);
      var theta = pcN * (lon - centralMeridian);

      var x = (r * Math.sin(theta) + falseEasting) * unitsPerMeter;
      var y = ((pcR0 - r * Math.cos(theta)) + falseNorthing) * unitsPerMeter;

      return [x, y];
    }

    function toGeodetic(p) {
      p[0] = (p[0] - falseEasting) * metersPerUnit;
      p[1] = (p[1] - falseNorthing) * metersPerUnit;

      var r = Math.sqrt(p[0] * p[0] + Math.pow(pcR0 - p[1], 2));
      var theta;

      if (pcN >= 0) {
        theta = Math.atan2(p[0], pcR0 - p[1]);
      }
      else {
        r = -r;
        theta = Math.atan2(-p[0], p[1] - pcR0);
      }

      var t = Math.pow(r / (spheroid.semiMajorAxis * pcF), 1 / pcN);

      var halfE = spheroid.eccentricity / 2;

      var phi;
      var newPhi = halfPi - 2 * Math.atan(t);
      var i = 0;

      do {
        phi = newPhi;
        var eSinPhi = spheroid.eccentricity * Math.sin(phi);
        newPhi = halfPi - 2 * Math.atan(t * Math.pow((1 - eSinPhi) / (1 + eSinPhi), halfE));
      }
      while (newPhi != phi && ++i < 10);

      var lambda = theta / pcN + centralMeridian;

      var lon = lambda * degreesPerRadian;
      var lat = phi * degreesPerRadian;

      return [lon, lat];
    }

    var lcc = {
      toProjected: toProjected,
      toGeodetic: toGeodetic,
      toUnits: function (meters) { return meters * unitsPerMeter; }
    };

    return lcc;
  };

  return gpv;
})(GPV || {});

File: WebApp/App_Code/Projection/LambertConformalConic.cs
//  Copyright 2012 Applied Geographics, Inc.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

using System;

public class LambertConformalConic : Projection
{
	// coordinate system

	private double _centralMeridian;
	private double _originLatitude;
	private double _standardParallel1;
	private double _standardParallel2;

	private Spheroid _spheroid;

	// projection constants

	private double _pcN;
	private double _pcF;
	private double _pcR0;

	public LambertConformalConic(double centralMeridian, double originLatitude,
			double standardParallel1, double standardParallel2, Spheroid spheroid)
	{
		_centralMeridian = centralMeridian * RadiansPerDegree;
		_originLatitude = originLatitude * RadiansPerDegree;
		_standardParallel1 = standardParallel1 * RadiansPerDegree;
		_standardParallel2 = standardParallel2 * RadiansPerDegree;
		_spheroid = spheroid;

		Initialize();
	}

	private void Initialize()
	{
		double m1 = GetM(_standardParallel1);
		double m2 = GetM(_standardParallel2);

		double t0 = GetT(_originLatitude);
		double t1 = GetT(_standardParallel1);
		double t2 = GetT(_standardParallel2);

		_pcN = (Math.Log(m1) - Math.Log(m2)) / (Math.Log(t1) - Math.Log(t2));
		_pcF = m1 / (_pcN * Math.Pow(t1, _pcN));
		_pcR0 = GetR(t0);
	}

	public override void ToGeodetic(double x, double y, out double lon, out double lat)
	{
		double dy = _pcR0 - y;
		double r = Math.Sqrt(x * x + dy * dy);
		double theta;

		if (_pcN >= 0)
		{
			theta = Math.Atan2(x, _pcR0 - y);
		}
		else
		{
			r = -r;
			theta = Math.Atan2(-x, y - _pcR0);
		}

		double t = Math.Pow(r / (_spheroid.SemiMajorAxis * _pcF), 1 / _pcN);

		double halfE = _spheroid.Eccentricity / 2;

		double phi;
		double newPhi = HalfPi - 2 * Math.Atan(t);
		int i = 0;

		do
		{
			phi = newPhi;
			double eSinPhi = _spheroid.Eccentricity * Math.Sin(phi);
			newPhi = HalfPi - 2 * Math.Atan(t * Math.Pow((1 - eSinPhi) / (1 + eSinPhi), halfE));
		}
		while (newPhi != phi && ++i < 10);

		double lambda = theta / _pcN + _centralMeridian;

		lon = lambda / RadiansPerDegree;
		lat = phi / RadiansPerDegree;
	}

	public override void ToProjected(double lon, double lat, out double x, out double y)
	{
		lon *= RadiansPerDegree;
		lat *= RadiansPerDegree;

		double t = GetT(lat);
		double r = GetR(t);
		double theta = _pcN * (lon - _centralMeridian);

		x = (r * Math.Sin(theta));
		y = (_pcR0 - r * Math.Cos(theta));
	}

  public override string ToProj4String()
  {
    return String.Format("+proj=lcc +lon_0={0} +lat_0={1} +lat_1={2} +lat_2={3} +ellps=GRS80 +datum=NAD83 +towgs84=0,0,0,0,0,0,0 +no_defs",
      _centralMeridian * DegreesPerRadian, _originLatitude * DegreesPerRadian, _standardParallel1 * DegreesPerRadian,
      _standardParallel2 * DegreesPerRadian);
  }

	private double GetM(double latitude)
	{
		double eSqr = _spheroid.Eccentricity * _spheroid.Eccentricity;
		return Math.Cos(latitude) / Math.Sqrt(1 - (eSqr * Math.Pow(Math.Sin(latitude), 2)));
	}

	private double GetR(double t)
	{
		return _spheroid.SemiMajorAxis * _pcF * Math.Pow(t, _pcN);
	}

	private double GetT(double latitude)
	{
		double sinLat = Math.Sin(latitude);
		return Math.Tan(QuarterPi - latitude / 2) /
			  Math.Pow((1 - _spheroid.Eccentricity * sinLat) /
			  (1 + _spheroid.Eccentricity * sinLat), _spheroid.Eccentricity / 2);
	}
}

